flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/sink/KafkaProducerRecordSerializationSchema.java

```diff
@@ -0,0 +1,64 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.streaming.connectors.kafka.sink;
+
+import org.apache.flink.annotation.PublicEvolving;
+import org.apache.flink.api.common.serialization.SerializationSchema;
+
+import org.apache.kafka.clients.producer.ProducerRecord;
+
+import java.io.Serializable;
+
+/**
+ * A serialization schema which defines how to convert a value of type {@code T} to {@link
+ * ProducerRecord}.
+ *
+ * @param <T> the type of values being serialized
+ */
+@PublicEvolving
+public interface KafkaProducerRecordSerializationSchema<T> extends Serializable {
+
+    /**
+     * Initialization method for the schema. It is called before the actual working methods {@link
+     * #serialize(Object, FlinkKafkaSinkContext, Long)} and thus suitable for one time setup work.
+     *
+     * <p>The provided {@link SerializationSchema.InitializationContext} can be used to access
+     * additional features such as e.g. registering user metrics.
+     *
+     * @param context Contextual information that can be used during initialization.
+     */
+    default void open(SerializationSchema.InitializationContext context) throws Exception {}
+
+    /**
+     * Serializes given element and returns it as a {@link ProducerRecord}.
+     *
+     * @param element element to be serialized
+     * @param context context to possibly determine target partition
+     * @param timestamp timestamp
+     * @return Kafka {@link ProducerRecord}
+     */
+    ProducerRecord<byte[], byte[]> serialize(
```

Comment:
The timestamp is already provided by another context [1] that is passed to `SinkWriter#write`. Merging it to the `FlinkKafkaSinkContext` IMO does not really make sense because currently the context is mainly used to let the `KafkaWriter` access information that is only present in the InitContext [2].

[1] https://github.com/apache/flink/blob/488e037285565cc5965c5130ecc92d8a727a4866/flink-core/src/main/java/org/apache/flink/api/connector/sink/SinkWriter.java#L67
[2] https://github.com/apache/flink/blob/488e037285565cc5965c5130ecc92d8a727a4866/flink-core/src/main/java/org/apache/flink/api/connector/sink/Sink.java#L85

Reply:
I don't quite get your arguments. My main idea: you don't need timestamp in 99% of the time (it's usually part of the payload). The old interface didn't even have the parameter. If there is still a need to have that information, the user can access it through `FlinkKafkaSinkContext` that contains all metadata about Kafka and the record.

Reply:
I think the old interface had this parameter [1]. Currently I also do not see a nice way how to pass the timestamp query to the context because it may also be a new context during `write` [2]

[1] https://github.com/fapaul/flink/blob/a6dcb7e41155cb5c209717d0fb7d9c76fa4b61a9/flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/KafkaSerializationSchema.java#L60
[2] https://github.com/fapaul/flink/blob/a6dcb7e41155cb5c209717d0fb7d9c76fa4b61a9/flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/sink/KafkaWriter.java#L93